core/src/main/java/com/linecorp/armeria/common/stream/FilteredStreamMessage.java

```diff
@@ -298,17 +219,21 @@ public void onNext(T o) {
             try {
                 filtered = filter(o);
             } catch (Throwable ex) {
-                StreamMessageUtil.closeOrAbort(o);
                 // onError(ex) should be called before upstream.cancel() to deliver the cause to downstream.
-                // upstream.cancel() and make downstream closed with CancelledSubscriptionException
-                // before sending the actual cause.
+                // upstream.cancel() may close downstream with CancelledSubscriptionException before sending
+                // the actual cause.
                 onError(ex);
 
                 assert upstream != null;
                 upstream.cancel();
                 return;
             }
 
+            if (completed) {
+                // onError(Throwable) or onComplete() has been called in filter().
+                StreamMessageUtil.closeOrAbort(filtered);
+                return;
```

Comment:
I didn't understånd this change 😅   Shouldn't the last filtered object be also passed downstream?

Reply:
We should not publish items if after calling `onError()` or onComplete()`.
https://github.com/reactive-streams/reactive-streams-jvm#1.7

Otherwise, did you mean something else?

Reply:
I imagined `FilteredStreamMessage#beforeComplete` publishing an item via:

https://github.com/line/armeria/blob/6fae642f27e25a3dc276690a3bc91b718fd65441/core/src/main/java/com/linecorp/armeria/common/stream/FilteredStreamMessage.java#L268-L270

where `completed=true` is set but `onNext` is still called in the following scenario:

https://github.com/line/armeria/blob/6fae642f27e25a3dc276690a3bc91b718fd65441/core/src/main/java/com/linecorp/armeria/server/encoding/HttpEncodedResponse.java#L163

Reply:
I fixed this to pass `StreamMessageCollectingTest.filteredStreamMessage_cancel()`.
https://github.com/line/armeria/blob/c7aca109dda4bedec8a84f96abe6c471aff05a1e/core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageCollectingTest.java#L185-L190

`subscription.cancel()` made upstream signal `subscriber.onError(CancelledSubscriptionException)` and then `subscribe.onNext()` was called.

It's impossible to know what all subclasses do. So we may need to prevent `onNext()` from being called by `beforeComplete()` or `beforeOnError()`.

```java
@Override
public void onNext(T o) {
    if (complete) {
        return;
    }

    U filtered;
    try {
        filtered = filter(o);
}
```

Reply:
By the way, only the `Subscription` API of `FilteringSubscriber` is exposed to the subclasses. It seems `FilteringSubscriber.onNext()` couldn't be called by `beforeCompete()`

Reply:
> By the way, only the Subscription API of FilteringSubscriber is exposed to the subclasses. 

I just realized that the downstream `Subscriber#onNext` is called instead of `FilteringSubscriber#onNext`. I guess the current implementation should be fine then 👍